src/test_fixture/world.rs

```diff
@@ -245,72 +252,35 @@ where
         <[_; 3]>::try_from(output).unwrap()
     }
 
-    async fn malicious<'a, O, M, H, R, P>(&'a self, input: I, mut helper_fn: H) -> [O; 3]
+    async fn malicious<'a, O, M, H, R, P>(&'a self, input: I, helper_fn: H) -> [O; 3]
     where
         A: IntoMalicious<F, M>,
         O: Send + Debug,
         M: Send,
-        H: FnMut(MaliciousContext<'a, F>, M) -> R + Send,
+        H: Fn(MaliciousContext<'a, F>, M) -> R + Send + Sync,
         R: Future<Output = P> + Send,
-        P: DowngradeMalicious<Target = O> + Send + Debug,
+        P: DowngradeMalicious<Target = O> + Clone + Send + Debug,
         [P; 3]: ValidateMalicious<F>,
         Standard: Distribution<F>,
     {
-        // The following is what this *should* look like,
-        // but so far the spelling necessary to convince the borrow checker
-        // to accept this has not been found.
-        //
-        // Current theory is that this might allow `helper_fn` to be run
-        // on multiple different threads concurrently, which would be bad.
-        // The long form below ensures that it is only run on one thread,
-        // even if it might move (with `Send`) a few times before it runs.
-
-        #[cfg(exemplary_code)]
-        {
+        let (m_results, r_shares, output) = split_array_of_tuples(
             self.semi_honest(input, |ctx, share| async {
                 let v = MaliciousValidator::new(ctx);
                 let m_share = share.upgrade(v.context()).await;
-                let res = helper_fn(v.context(), m_share).await;
-                v.validate(res).await.unwrap()
+                let m_result = helper_fn(v.context(), m_share).await;
+                let m_result_clone = m_result.clone();
+                let r_share = v.r_share().clone();
+                let output = v.validate(m_result_clone).await.unwrap();
```

Comment:
```suggestion
                let r_share = v.r_share().clone();
                let output = v.validate(m_result.clone()).await.unwrap();
```

Reply:
... and now I see why you needed to do this explicitly.  That `await` really messes with things.